### ISSUE TITLE
Clearly state plans for stencil as an standalone framework, or a deprecation timeline

### ISSUE THREAD

**Issue by adrm** [ionic-team/stencil#3071]
### Prequisites

- [X] I have read the [Contributing Guidelines](https://github.com/ionic-team/stencil/blob/master/.github/CONTRIBUTING.md).
- [X] I agree to follow the [Code of Conduct](https://github.com/ionic-team/stencil/blob/master/CODE_OF_CONDUCT.md).
- [X] I have searched for [existing issues](https://github.com/ionic-team/stencil/issues) that already include this feature request, without success.

### Describe the Feature Request

I asked in ionic about support for stencil as a framework, without embedding its components in other framework. Liam answered that you are right now stating in your FAQs that Stencil is no a stand-alone framework, and that is a bit of a change of course for the project.

We are right now relying on Stencil as an stand-alone framework, since this seemed as a secondary objective for the project. Everything points right now to a change in direction, so I wonder if you are planning to phase out this stencil use-case. I imagine the answer is yes, but I would love to get a clear answer, and a deprecation timeline for features as service worker support, stencil router, stencil store and other features that only make sense in a stand-alone use case.

### Describe the Use Case

We are using Stencil as an stand-alone framework but it seems this use case is not going to be maintained.

### Describe Preferred Solution

State an official course of action about this, and a deprecation timeline if the decision is, as it seems, to phase out the stand-alone framework use case. This would help those of us using Stencil this way to plan accordingly and be prepared.

**Comment by splitinfinities** [comment]
Hey @adrm, thanks for reaching out about this! I want to be quite clear, not only because 1. this distinction is important to the team & product strategy 2. because I’m now the PM for Stencil (for now, I split my time 40% PM and 60% Engineering) and 3. I want folks in the community to understand the goals of Stencil as a product. 

**TL;DR**: We are **_not_** deprecating support for your web app "framework" use case. We’re actively improving Stencil, and will continue to find and incorporate ways to improve _all_ component use cases. 

I want to share the philosophy that generally guides the Stencil team, which I share at length only so you - and the many others inevitably reading this comment - can be more at ease about Stencil and its future. We are a new team, which I think can be scary for community members who are using Stencil. So, yeah, I think it’s totally worth sharing where we’re at today and really just reinforcing a few perspectives that have been shared since the Stencil came into existence from a brand new face who's now working full time on the team.  

I want you to feel comfortable to continue to use Stencil as a framework for your apps - we will not be removing that use case. What's crucial to share here is that the Stencil team doesn't see Stencil as a framework. Stencil helps people compile components by way of web components, paired up with a JSX variant for templating. We’re not changing that any time soon (But read on for more potential product strategy around this). Stencil is not being deprecated either. As a matter of fact, we’re actively improving the product’s rough edges on a daily basis. 

Frankly, the team is fascinated by _**all**_ the use cases of the wider concept of “Components”.  No matter if you use a component as a full app (collections of page components), a page (collections of section/block components), a section/block (collections of design system components), or the smallest component of your Design System - you’ll have support from Stencil in compiling components, period. This seems similar to React and Vue, as frameworks, right? Almost indistinctly. The crucial difference here is that our product strategy around compiling web components is to compile as much “Stencil” stuff away and rely on widely available platform features. 

Strategically, we would love to improve ways to help guide your use case for building web apps as components. A better way to say that: Stencil helps folks realize a breadth of technical strategies - from single component component libraries, to any-sized Design Systems, to full on multi-platform PWAs with Design Systems backing them. 

Specifically around the web app use case: In the future, I’d like to explore something like putting the PWA features into a separate peer dependency like “@stencil/pwa” that would be prompted to you as a user to be installed only when it becomes obvious that you need those features. That’s how Jest and Puppeteer work today; you get a prompt to install other tangentially related things before the `npx stencil test` task runs. Introducing this 1. may help reduce the complexity of Stencil’s codebase 2. may make Stencil's compilation process even more effective (aka one of Stencil's few core features) 3. We may be able to give that segment of Stencil’s offerings more love as well and 4. Generally may give us more velocity to ship features and fix bugs as we can more quickly determine the location of different concepts inside of Stencil code. FYI: **_Nothing_** on the roadmap here, I’m only attempting to share our philosophy and how it could be put into practice. 

Here’s something else that fits with the “Remove all Stencil stuff from the final compile” philosophy. Maybe as browsers add support for Shadow DOM to adopt parent styles into the shadow root like in https://github.com/WICG/webcomponents/issues/909, we would consider switching our scoped styles feature to now use that behavior instead. That removes as much "Stencil" stuff as possible in the final compiled files. Especially as IE11 begins to reach its EOL. **_Nothing_** on the roadmap here, just wanted to share how that philosophy can be put into practice. 

One final example of how our philosophy could be put into practice: Templating. There are indeed folks out there who do not like JSX. However, until we see a viable solution widely built into browsers, we are going to continue to use JSX. It may very well seem to you as though Stencil is a Framework similar to React, which is often misinterpreted because we use JSX - but we have an explicit goal of actually wanting to remove as much “Stencil” code as possible in the final compiled output. Again, **_nothing_** detailed in this paragraph is on the roadmap/planned. 

I have many more ways that this philosophy can be put to practice (*cough cough* Declarative Shadow DOM *cough*) that I've been excitedly prototyping. 

We have a deeply held belief that the platform is capable of solving problems that our users have, and we want to compile components independent of the level where/how the component is used. That’s a part of the reason we say we’re not a framework; from our perspective the less Stencil “stuff” in the end result of a compile the better. I hope I do sound like a broken record about this too. Stencil is a compiler that creates components. You need React and Vue runtime to run React and Vue apps, and we don’t want you to _**need**_ Stencil to run apps compiled by Stencil. That distinction is crucial to Stencil as a product, and one we need to do an increasingly better job at communicating to folks. Hopefully this response helps. This is where all the “We’re not a framework” language emerges, and it’s what we believe sets Stencil apart from a traditional JS "Framework". 

Please let me know your feedback! I hope this brought clarity, like you're asking for. I really do want folks to understand Stencil's product strategy. Especially since we are a new team working on Stencil. I'll keep this issue open until Friday then close it. 

With all that said, we're also looking to hire folks on the Stencil team. If anyone wants to join up and realize this philosophy—and many, many other cool and exciting framework integrations—please apply!

https://jobs.lever.co/Ionic/59e7e6eb-58d7-4ddb-a51d-ee5ffa40abef

PS: Personal note that is not of the Ionic/Stencil opinion: In some ways, I feel like Stencil's product strategy aligns less to React or Vue product strategies, and more toward Eleventy's product strategy of being a compiler for HTML. Eleventy is not a framework - and no one AFAIK ever calls it a framework - it's a compiler/generator. We do have many more opinions within Stencil, but spiritually the product strategy is pretty gosh dang similar. At the core of what we do, we really do want to remove as much "Stencil" stuff in the final compile. That's kind of where Stencil's "identity crisis" is at today, which falls predominately on my shoulders to work through. When I read your comment last week, it got me thinking about this. So the goal of my response here today is to set the ground work to solve that identity crisis and give everyone clarity. Don't worry - much more to come!

**Comment by MarkChrisLevy** [comment]
@splitinfinities Great explanation, thanks for that! I fully agree with your strategy and this is exactly how I see and use Stencil - not a framework, but rather a tool/compiler/generator to produce components that heavily relies on platform features. In my case Stencil is everywhere: I make mobile full apps with it (ionic + capacitor), I make standalone web components and finally I make website/pwa (ssr).

What would be great is to have a roadmap, where all planned changes, improvements, feature request will be listed (something like typescript's roadmap).

Fingers crossed for the new team!

**Comment by splitinfinities** [comment]
Finally, another point to drive home: the router, store, state-tunnel, and a few others. I shared in Slack last week that I had worked through the level of support that each of our npm packages & GitHub repos will  receive with our current team size. We are looking for more developers, and I'm trying to determine ways to improve how we can support these repos. There are a few decisions I will be able to share sooner than later once we are all bought in, I just need to find more time to do more research to justify those decisions - e.g. I'm thinking Router v1 versus v2, which one do we officially support and why? Those are questions I'm trying to answer ASAP and once I have answered, I'll share with the community and I'll try to organize an upgrade path so it can be painless.

I hope that also helps!

**Comment by adrm** [comment]
Thanks for being straightforward about your position right now. I understand that you are a new team, taking the reins of this project, so not everything is absolutely clear and that is fine.

One of the things that made us choose Stencil in the first place is this "not a framework, but a compiler" approach, so it is good to see this is here to stay.

We were concerned about the use of Stencil for a full app, instead of using it as a compiler for components that are to be used in other projects with more usual frameworks, and we still are a bit. I appreciate your interest in listening to the community to decide how to proceed. But I also understand that the incentives for Stencil inside Ionic the company are more focused towards design systems and component libraries. Furthermore, as I told before, the Ionic Framework team stated they will not support the Stencil + Ionic use case specifically, but thankfully, being standard web components, almost everything will work out of the box.

The proposal of a new `@stencil/pwa` package could be a great approach. Maybe that package could provide a set of Rollup plugins that can allow to have a native Rollup config (not through the Stencil compiler) that compiles Stencil components, configures service worker support... Thus decoupling PWA support from the core compiler, and decoupling Rollup config for a full PWA from the Stencil compiler.

Anyway, we'll keep betting on Stencil for the time being, as it has served us well and your enthusiasm for the project looks very promising for its future.

Good luck and thanks a lot!

**Comment by splitinfinities** [comment]
Thanks all for being so understanding! I really appreciate the perspective. Have a great weekend!

**Comment by cjorasch** [comment]
One thing that is important to keep in mind is that the use of Stencil as a framework significantly reduces the learning curve when creating applications.  It is awkward to write components in Stencil and then use React for the remainder of the app - especially since there is not a firm line between what should be a component and what should be in the app.  Often things might start out in app code and later be converted to shared components.

We have found pure Stencil with Ionic to be a very good framework and most of the limitations are related to things like routing and nav that could be improved in Ionic.

A key for Stencil is to ensure a rock solid foundation for components and apps and it would be great to see some of the long standing foundational issues get resolved.
